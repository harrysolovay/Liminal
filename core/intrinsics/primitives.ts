import { assert } from "@std/assert"
import type { Type } from "../Type.ts"
import { declare } from "./declare.ts"

export { null_ as null }
const null_: Type<null> = declare({
  getAtom: () => null_,
  assert: (value) => {
    assert(value === null)
  },
})

export const boolean: Type<boolean> = declare({
  getAtom: () => boolean,
  assert: (value) => {
    assert(typeof value === "boolean")
  },
})

export const integer: Type<number> = declare({
  getAtom: () => integer,
  assert: (value) => {
    assert(Number.isInteger(value))
  },
})

export const number: Type<number> = declare({
  getAtom: () => number,
  assert: (value) => {
    assert(typeof value === "number")
  },
})

export const string: Type<string> = declare({
  getAtom: () => string,
  assert: (value) => {
    assert(typeof value === "string")
  },
})
