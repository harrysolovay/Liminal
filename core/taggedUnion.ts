import type { Type } from "../Type.ts"
import { declare } from "../TypeDeclaration.ts"
import type { Expand } from "../util/type_util.ts"

// TODO: allow numbers as tags + tag keys?
export function taggedUnion<
  K extends string,
  M extends Expand<Record<string, Type | null>>,
>(
  tagKey: K,
  members: M,
): Type<
  { [V in keyof M]: ({ [_ in K]: V } & (M[V] extends Type ? { value: M[V]["T"] } : {})) }[keyof M],
  {},
  Extract<M[keyof M], Type>["P"]
> {
  const entries = Object.entries(members)
  return declare({
    name: "taggedUnion",
    source: {
      factory: taggedUnion,
      args: { tagKey, members },
    },
    subschema: (visit) => ({
      discriminator: tagKey,
      anyOf: entries.map(([k, v]) => ({
        type: "object",
        properties: {
          [tagKey]: {
            type: "string",
            const: k,
          },
          ...(v === null ? {} : { value: visit(v) }),
        },
        required: [tagKey, ...v === null ? [] : ["value"]],
        additionalProperties: false,
      })),
    }),
    output: (f) =>
      f<{ [V in keyof M]: ({ [_ in K]: V } & { value?: unknown }) }[keyof M]>({
        visitor: (value, visit, path) => {
          const tag = value[tagKey]
          const type = members[tag]!
          return ({
            [tagKey]: value[tagKey],
            // TODO: fix typing of `tag`.
            ..."value" in value
              ? {
                value: visit(
                  value.value,
                  type,
                  path.value("value").type(tag as string),
                ),
              }
              : {},
          }) as never
        },
      }),
  })
}
