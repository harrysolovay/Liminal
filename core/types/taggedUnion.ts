import type { Expand } from "../../util/mod.ts"
import { type AnyType, Type } from "../Type.ts"

export function taggedUnion<
  K extends number | string,
  M extends Record<number | string, AnyType | undefined>,
>(
  tagKey: K,
  members: M,
): Type<
  {
    [V in keyof M]: Expand<
      (
        & { [_ in K]: V }
        & (M[V] extends AnyType ? { value: M[V]["T"] } : {})
      )
    >
  }[keyof M],
  Extract<M[keyof M], AnyType>["P"]
> {
  return Type({
    kind: "taggedUnion",
    factory: taggedUnion,
    args: [tagKey, members],
    argsLookup: { tagKey, members },
  })
}
