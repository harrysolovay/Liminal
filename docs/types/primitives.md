# Primitive Types

```ts twoslash
import { L } from "liminal"
// ---cut---
L.boolean
// ^?
```

<br />

```ts twoslash
import { L } from "liminal"
// ---cut---
L.number
// ^?
```

<br />

```ts twoslash
import { L } from "liminal"
// ---cut---
L.integer
// ^?
```

<br />

```ts twoslash
import { L } from "liminal"
// ---cut---
L.string
// ^?
```

<br />
