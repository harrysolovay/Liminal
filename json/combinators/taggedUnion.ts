import { type AnyType, Type } from "../../core/mod.ts"
import type { Expand } from "../../util/type_util.ts"

export function taggedUnion<
  K extends number | string,
  M extends Record<number | string, AnyType | null>,
>(
  tagKey: K,
  members: M,
): Type<
  {
    [V in keyof M]: Expand<({ [_ in K]: V } & (M[V] extends AnyType ? { value: M[V]["T"] } : {}))>
  }[keyof M],
  Extract<M[keyof M], AnyType>["P"]
> {
  return Type({
    factory: taggedUnion,
    args: [tagKey, members],
  })
}
