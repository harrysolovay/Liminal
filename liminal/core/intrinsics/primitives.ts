import { assert } from "@std/assert"
import { declare } from "../declare.ts"
import type { Type } from "../Type.ts"

export { null_ as null }
const null_: Type<null, never> = declare({
  jsonType: "null",
  getAtom: () => null_,
  assert: (value) => {
    assert(value === null)
  },
})

export const boolean: Type<boolean, never> = declare({
  jsonType: "boolean",
  getAtom: () => boolean,
  assert: (value) => {
    assert(typeof value === "boolean")
  },
})

export const integer: Type<number, never> = declare({
  jsonType: "integer",
  getAtom: () => integer,
  assert: (value) => {
    assert(Number.isInteger(value))
  },
})

export const number: Type<number, never> = declare({
  jsonType: "number",
  getAtom: () => number,
  assert: (value) => {
    assert(typeof value === "number")
  },
})

export const string: Type<string, never> = declare({
  jsonType: "string",
  getAtom: () => string,
  assert: (value) => {
    assert(typeof value === "string")
  },
})
