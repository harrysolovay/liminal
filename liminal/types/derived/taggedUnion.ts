import type { JSONObjectType } from "../_object.ts"
import type { JSONConstType } from "../const.ts"
import type { JSONStringType } from "../string.ts"
import { type Type } from "../Type.ts"
import { type JSONUnionType, union } from "../union.ts"
import { type NormalizeObjectValueJ, type NormalizeObjectValueT, object, type ObjectFieldValue } from "./object.ts"

export function taggedUnion<const M extends TaggedUnionMembers, K extends Extract<keyof M, string>>(members: M): Type<
  {
    [L in K]: {
      key: L
      value: NormalizeObjectValueT<M[L]>
    }
  }[K],
  JSONUnionType<
    {
      [L in K]: JSONObjectType<{
        key: JSONConstType<JSONStringType, L>
        value: NormalizeObjectValueJ<M[L]>
      }>
    }[K]
  >
> {
  return union(
    ...Object.entries(members).map(([key, value]) => object({ key, value })),
  ) as never
}

export type TaggedUnionMembers = Record<string, ObjectFieldValue>
