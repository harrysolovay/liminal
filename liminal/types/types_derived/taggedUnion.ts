import type { JSONKey } from "../../util/JSONKey.ts"
import type { JSONConstType } from "../const.ts"
import { type JSONObjectType, object } from "../object.ts"
import type { JSONStringType } from "../string.ts"
import { type Type } from "../Type.ts"
import { type NormalizeTypeLikeJ, type NormalizeTypeLikeT, type TypeLike } from "../TypeLike.ts"
import { _union, type JSONUnionType, union } from "../union.ts"

export function taggedUnion<const M extends TaggedUnionMembers, K extends Extract<keyof M, string>>(members: M): Type<
  {
    [L in K]: {
      type: L
      value: NormalizeTypeLikeT<M[L]>
    }
  }[K],
  JSONUnionType<
    {
      [L in K]: JSONObjectType<{
        type: JSONConstType<JSONStringType, L>
        value: NormalizeTypeLikeJ<M[L]>
      }>
    }[K]
  >
> {
  return union(
    ...Object.entries(members).map(([type, value]) => object({ type, value })),
  ) as never
}

export type TaggedUnionMembers = Record<JSONKey, TypeLike>
