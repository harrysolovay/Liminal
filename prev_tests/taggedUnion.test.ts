import { assertTySnapshot } from "test_util"
import { boolean, number, string } from "./primitives.ts"
import { taggedUnion } from "./taggedUnion.ts"

Deno.test("tagged union", async (t) => {
  await assertTySnapshot(
    t,
    taggedUnion("type", {
      A: boolean,
      B: number,
      C: string,
    }),
  )
})
