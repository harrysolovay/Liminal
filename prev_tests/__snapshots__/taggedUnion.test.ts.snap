export const snapshot = {};

snapshot[`tagged union 1`] = `
{
  anyOf: [
    {
      additionalProperties: false,
      properties: {
        type: {
          const: "A",
          type: "string",
        },
        value: {
          type: "boolean",
        },
      },
      required: [
        "type",
        "value",
      ],
      type: "object",
    },
    {
      additionalProperties: false,
      properties: {
        type: {
          const: "B",
          type: "string",
        },
        value: {
          type: "number",
        },
      },
      required: [
        "type",
        "value",
      ],
      type: "object",
    },
    {
      additionalProperties: false,
      properties: {
        type: {
          const: "C",
          type: "string",
        },
        value: {
          type: "string",
        },
      },
      required: [
        "type",
        "value",
      ],
      type: "object",
    },
  ],
  discriminator: "type",
}
`;

snapshot[`tagged union 2`] = `
{
  anyOf: [
    {
      additionalProperties: false,
      properties: {
        type: {
          const: "A",
          type: "string",
        },
        value: {
          type: "boolean",
        },
      },
      required: [
        "type",
        "value",
      ],
      type: "object",
    },
    {
      additionalProperties: false,
      properties: {
        type: {
          const: "B",
          type: "string",
        },
        value: {
          type: "number",
        },
      },
      required: [
        "type",
        "value",
      ],
      type: "object",
    },
    {
      additionalProperties: false,
      properties: {
        type: {
          const: "C",
          type: "string",
        },
        value: {
          type: "string",
        },
      },
      required: [
        "type",
        "value",
      ],
      type: "object",
    },
  ],
  description: "One.",
  discriminator: "type",
}
`;

snapshot[`tagged union 3`] = `
{
  anyOf: [
    {
      additionalProperties: false,
      properties: {
        type: {
          const: "A",
          type: "string",
        },
        value: {
          type: "boolean",
        },
      },
      required: [
        "type",
        "value",
      ],
      type: "object",
    },
    {
      additionalProperties: false,
      properties: {
        type: {
          const: "B",
          type: "string",
        },
        value: {
          type: "number",
        },
      },
      required: [
        "type",
        "value",
      ],
      type: "object",
    },
    {
      additionalProperties: false,
      properties: {
        type: {
          const: "C",
          type: "string",
        },
        value: {
          type: "string",
        },
      },
      required: [
        "type",
        "value",
      ],
      type: "object",
    },
  ],
  description: "Two. One.",
  discriminator: "type",
}
`;
